impl/deps/postgres/instant/instant.go

```diff
@@ -0,0 +1,218 @@
+	id := uuid.New().String()
+
+	tx, err := d.client.Begin(ctx)
+	if err != nil {
+		return "", fmt.Errorf("failed to begin a postgres transaction: %s", tx)
+	}
+	defer tx.Rollback(ctx)
+
+	sql := `INSERT INTO InstantGroups (id, pending)
+			VALUES ($1,
+			        $2 - (SELECT COUNT(*) FROM InstantDependencies d WHERE d.resolved)
+			)`
+	_, err = tx.Exec(ctx, sql, id, len(ids))
+	if err != nil {
+		return "", err
+	}
+
+	for _, dep := range ids {
+		sql := `INSERT INTO InstantGroupDependencies (dependency_id, group_id) VALUES ($1, $2)`
+		_, err := tx.Exec(ctx, sql, dep, id)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	if err := tx.Commit(ctx); err != nil {
+		return "", fmt.Errorf("failed to commit the transaction: %s", err)
+	}
+
+	return id, nil
+}
+
+func (d *Deps) Resolve(ctx context.Context, dep string) error {
+	//if err := d.resolvedDeps.Write(ctx, dep); err != nil {
+	//	return fmt.Errorf("failed to write dep to queue: %s", err)
+	//}
+	//return nil
+
+	tx, err := d.client.Begin(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to begin a postgres transaction: %s", tx)
+	}
+	defer tx.Rollback(ctx)
+	// let's goooooo
+	// 1) update dep and check if it has been resolved
+	sql := `
+		WITH previous AS (
+			SELECT resolved
+			FROM InstantDependencies
+			WHERE id = $1
+			LIMIT 1
+		)
+		INSERT INTO InstantDependencies (id, resolved) 
+		VALUES ($1, $2)
+		ON CONFLICT (id) DO UPDATE 
+		SET resolved = $2
+		RETURNING COALESCE((SELECT resolved FROM previous), false) AS previous_resolved;`
+	before := tx.QueryRow(ctx, sql, dep, true)
+	var alreadyResolved bool
+	if err := before.Scan(&alreadyResolved); err != nil {
+		return err
+	}
+	if alreadyResolved {
+		// if it resolved we are done
+		return nil
+	}
+
+	// dep hasn't been resolved, so we need to update groups
+	// we also return groups which have no unresolved deps
+	sql = `
+			UPDATE InstantGroups g
+			SET pending = pending - 1
+			FROM InstantGroupDependencies gd
+			WHERE gd.dependency_id = $1 AND gd.group_id = g.id
+			RETURNING (SELECT g.id WHERE g.pending = 0)` // returns new value
+	resolved, err := tx.Query(ctx, sql, dep)
+	var groups []string
+	for resolved.Next() {
+		// I do not understand wny it returns nulls when result should be empty
+		var id *string
+		if err := resolved.Scan(&id); err != nil {
+			return fmt.Errorf("failed to scan the id of a group (please report if it ever happens): %s", err)
+		}
+		if id != nil {
+			groups = append(groups, *id)
+		}
+	}
+	// now we just add groups to their queue
+	if err := d.resolvedGroups.Write(ctx, groups...); err != nil {
+		return fmt.Errorf("failed to add groups to queue")
+	}
+
+	if err := tx.Commit(ctx); err != nil {
+		return fmt.Errorf("failed to commit the transaction: %s", err)
+	}
+	return nil
+}
+
+func (d *Deps) Ready(ctx context.Context) (<-chan transactional.Object[string], error) {
+	return d.resolvedGroups.Read(ctx)
+}
+
+func (d *Deps) Run(_ context.Context) {}
```

Comment:
why do we need a `Run` function? (it's not in the interface)
more importantly why do we need an unused context argument in a non-interface function?

Reply:
there is an interface in postgres/common.go which allows you to run postgres deps of both types